Switch to yaml.safe_load(_all) to prevent YAMLLoadWarning

mbuijs opened the pull request:
Fixes #1686.

In https://github.com/yaml/pyyaml/issues/265 it's stated that `safe_load` always existed for this reason, I did a quick double check and it traces back at least to 2006, so this might even be safe to backport to older releases.

mbuijs committed:
Switch to yaml.safe_load(_all) to prevent YAMLLoadWarning

dirk-thomas commented:
Thanks for the patch. Can you maybe also update the other places in this repo where these functions are used?

mbuijs commented:
I can have a look. Would you prefer this also to be updated in tests?

dirk-thomas commented:
> Would you prefer this also to be updated in tests?

Yes, that would be good.

mbuijs committed:
Change all usages of yaml.load to yaml.safe_load

mbuijs commented:
So I had not realized about the following (from https://wiki.ros.org/rosparam):
> There are also special converters for angle radian/degree representations. Any Python-legal mathematical expression can be used with the radian value, with pi used to represent pi.

I'm doing some more changes to support this functionality while using the `SafeLoader` provided by pyyaml.

Furthermore I'll also try to see if there are other places where more features than the `SafeLoader` provides, so far I had just done a rapid search and replace...

mbuijs commented:
Additional question: which other tools aside of `rosparam` and `roslaunch` should have the ability to do the degrees and radians conversion functionality when evaluating YAML strings?

Edit: never mind, I see that also currently it is only supported in these 2, so I'll stick to that.

mbuijs committed:
Extend PyYAML's SafeLoader and use it with `yaml.load`

Also added convenience functions for using this loader for reuse in
`roslaunch`

mbuijs committed:
fix typo in rosparam.yaml_load_all

mbuijs commented:
So tests are passing now.
My experience with python is limited, I'm open to feedback, especially on how I applied the change in `rosparam` and `roslaunch`
1. attempting to not change the `SafeLoader` instance in PyYAML
2. adding the convencience functions `yaml_load` and `yaml_load_all` inside `rosparam`.

marting87 commented:
Changes look to me, since I'm getting a lot of **YAMLLoadWarning** in my console I hope to see this merged.

dirk-thomas commented on tools/roslaunch/src/roslaunch/loader.py:
```diff
@@ -406,11 +403,9 @@ def load_rosparam(self, context, ros_config, cmd, param, file_, text, verbose=Tr
-            global rosparam
-            if rosparam is None:
-                import rosparam
+            import rosparam
```
Why has this import logic been changed? It looks unrelated to the patch.

mbuijs replied:
You're right, it's unrelated to the functional change in this patch.

The reason I touched it was that I was checking if/where `rosparam` was imported within `roslaunch` globally and finding the global `rosparam` variable confused me. Then I googled for this construction and came across https://stackoverflow.com/a/4177780/177643, where it's pointed out that python already caches imports, so this optimization would not really help performance.
Aside of that, it pollutes the interface of this python file, by exposing `rosparam` which might be `None` or the loaded `rosparam` module, depending on if this function has been called or not. Since this is the only use of rosparam within this module, I thought it would be a good idea to clean up.

It's fine for me to roll this back, you can decide.

dirk-thomas replied:
Please remove the change from this PR since the title won't let you see that this is part of the patch. Please feel free tp create a separate PR with this change.

dirk-thomas commented on tools/rosparam/src/rosparam/__init__.py:
```diff
@@ -150,6 +151,12 @@ def print_params(params, ns):
+    return yaml.load(str, Loader=ROSSafeLoader)
+
+def yaml_load_all(str):
+    return yaml.load_all(str, Loader=ROSSafeLoader)
```
The naming of both functions is misleading since it doesn't mention that it is actually using a safe loader.

mbuijs replied:
As mentioned below I will remove these functions and instead add the resolvers to the `SafeLoader` object in the `yaml` module.

dirk-thomas commented on tools/rosparam/src/rosparam/__init__.py:
```diff
@@ -631,12 +638,12 @@ def yamlmain(argv=None):
+ROSSafeLoader.add_implicit_resolver(u'!degrees', pattern, first="deg(")
 pattern = re.compile(r'^rad\([^\)]*\)$')
-yaml.add_implicit_resolver(u'!radians', pattern, first="rad(")
+ROSSafeLoader.add_implicit_resolver(u'!radians', pattern, first="rad(")
```
Before this change all the added constructors and implicit resolvers affected the global yaml module. If any other existing code relies on that behavior it would break with this change.

mbuijs replied:
Good point, I'll adjust it to just modify the global `SafeLoader`. Then I can also remove the added yaml functions that I had added.

mbuijs committed:
Modify Loader and SafeLoader in yaml module directly

mbuijs commented:
@ros-pull-request-builder retest this please

mbuijs commented:
@dirk-thomas I processed your remarks and all checks have passed. Would you mind doing another check and merging if it's good?

dirk-thomas commented on tools/rosparam/src/rosparam/__init__.py:
```diff
@@ -149,7 +150,6 @@ def print_params(params, ns):
         print(params)
     
 # yaml processing
-
```
Please revert this unrelated white space change.

mbuijs committed:
Revert whitespace change

mbuijs committed:
Revert unrelated change to import through global variable construction

dirk-thomas commented:
Thank you for iterating on this.
